compiler/qsc_formatter/src/formatter/tests.rs

```diff
@@ -0,0 +1,307 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use expect_test::{expect, Expect};
+use indoc::indoc;
+
+fn check(input: &str, expect: &Expect) {
+    let actual = super::format(input);
+    expect.assert_debug_eq(&actual);
+}
+
+#[test]
+fn test_formatting() {
+    check(
+        "operation   Foo   ()",
+        &expect![[r#"
+    [
+        Edit {
+            span: Span {
+                lo: 9,
+                hi: 12,
+            },
+            new_text: " ",
+        },
+        Edit {
+            span: Span {
+                lo: 15,
+                hi: 18,
+            },
+            new_text: "",
+        },
+    ]
+"#]],
+    );
+}
+
+#[test]
+fn remove_trailing_spaces() {
+    let extra_spaces = "    ";
+    let input = format!(
+        "/// Doc Comment with trailing spaces{extra_spaces}
+operation Foo() : Unit {{
+    // Comment with trailing spaces{extra_spaces}
+    let x = 3;   // In-line comment with trailing spaces{extra_spaces}
+    let y = 4;{extra_spaces}
+}}
+"
+    );
+
+    check(
+        input.as_str(),
+        &expect![[r#"
+            [
+                Edit {
+                    span: Span {
+                        lo: 0,
+                        hi: 40,
+                    },
+                    new_text: "/// Doc Comment with trailing spaces",
+                },
+                Edit {
+                    span: Span {
+                        lo: 70,
+                        hi: 105,
+                    },
+                    new_text: "// Comment with trailing spaces",
+                },
+                Edit {
+                    span: Span {
+                        lo: 123,
+                        hi: 166,
+                    },
+                    new_text: "// In-line comment with trailing spaces",
+                },
+                Edit {
+                    span: Span {
+                        lo: 181,
+                        hi: 186,
+                    },
+                    new_text: "\n",
+                },
+            ]
+        "#]],
+    );
+}
+
+#[test]
+fn correct_indentation() {
```

Comment:
What kind of indentation is being corrected? What kinds of indents can be corrected? What test cases are we missing?

Reply:
This function is meant to ensure the given whitespace has the correct indentation and at least one newline. If the whitespace doesn't, it will generate an edit with the correct indentation and at least one newline.